### PR TITLE
PCT-625: Handle failure when number of agents exceeded

### PR DESCRIPTION
**Pull request by arvenil**
* PCT-625: Handle failure when number of agents exceeded
* Brought back to life all tests, plus extending them on e.g. checking if config files were properly created
* Move all api interactions to separate package and inject it as dependency
* Installer no longer pushes configs to api, this is done by agent when it connects to api (https://github.com/percona/cloud-api/blob/a3fcaf42980efb39468017d0a91c2e717f8abdc0/app/gatekeeper/agent.go#L14-L44).
  Reasons are:
  * This was actually broken - agent, log and data configs were not sent
  * There is no need to do that as agent sents them anyway when it connects
  * Single source of configs creation
  * KIS - simpler design and  it also allowed to simpler fix for PCT-625
* Resulted in https://github.com/percona/cloud-api/pull/308

### CODE REVIEW

File: bin/percona-agent-installer/main_test.go [hunk 1]
```diff
@@ -125,7 +128,7 @@ func (s *MainTestSuite) SetUpTest(t *C) {
 	t.Assert(err, IsNil)
 
 	// Remove config dir between tests.
-	err = os.RemoveAll(path.Join(s.basedir, pct.CONFIG_DIR))
+	err = os.RemoveAll(pct.Basedir.Path())
```
Comment (cfsalguero) [inline bin/percona-agent-installer/main_test.go]: There is an assert at line 128 that would prevent removing temp files

File: pct/basedir.go [hunk 2]
```diff
@@ -151,3 +151,15 @@ func (b *basedir) File(file string) string {
 	}
 	return filepath.Join(b.Path(), file)
 }
+
+func ReadConfig(configFile string, v interface{}) error {
+	data, err := ioutil.ReadFile(configFile)
+	if err != nil && !os.IsNotExist(err) {
```
Comment (cfsalguero) [inline pct/basedir.go]: Why checking the error is not IsNotExists if we are returning the error anyway?

Comment (arvenil) [inline pct/basedir.go]: Good question. I've only moved this code... @daniel-nichter ?

File: bin/percona-agent-installer/main_test.go (hunk 1 above)
Comment (arvenil) [inline bin/percona-agent-installer/main_test.go]: Thanks! Fixed!

File: pct/basedir.go (hunk 2 above)
Comment (daniel-nichter) [inline pct/basedir.go]: Idea was probably to ignore `IsNotExists` and just return no data. So maybe the final line of this func (`return err`) should be just `return nil`. See how callers use this func and what they do if file doesn't exist, how they handle errors and such.

File: test/installer/my.cnf-wrong_user [hunk 3]
```diff
@@ -1,4 +1,3 @@
 [client]
 user=wrong_user
-password=wrong_pass
```
Comment (daniel-nichter) [inline test/installer/my.cnf-wrong_user]: Why this change? Real .my.cnf usually have passwords in them.

Comment (arvenil) [inline test/installer/my.cnf-wrong_user]: Because there is a bug in installer that prevents setting password to `` (empty) with flag. And `my.cnf-wrong_user` wasn't used in anywhere so I claimed it for my test.

File: pct/basedir.go [hunk 4]
```diff
@@ -110,15 +118,7 @@ func (b *basedir) ConfigFile(service string) string {
...
-	data, err := ioutil.ReadFile(configFile)
-	if err != nil && !os.IsNotExist(err) {
-		// There's an error and it's not "file not found".
-		return err
-	}
-	if len(data) > 0 {
-		err = json.Unmarshal(data, &v)
-	}
-	return err
+	return ReadConfig(configFile, v)
```
Comment (daniel-nichter) [inline pct/basedir.go]: Is this necessary? It looks like it was done so that `main_test.go` can read config files, but that test already has [an instance Basedir](https://github.com/percona/percona-agent/blob/PCT-625_installer/bin/percona-agent-installer/main_test.go#L77-81), so it can just use the original instance method. Plus, for agent this encapsulation matters: it means agent must read all config files relative to basedir because there's no function to read arbitrary files (presuming caller doesn't use ioutils directly). I think with two `ReadConfig()` new/other devs will ask: "which one to use? when/why?"

File: pct/basedir.go (hunk 2 above)
Comment (arvenil) [inline pct/basedir.go]: @daniel-nichter I already checked that prior to asking and callers check for `os.IsNotExist(err)` so returning nil will brake them.

File: pct/basedir.go (hunk 4 above)
Comment (arvenil) [inline pct/basedir.go]: I will revert that. Test initially didn't had proper instance of Basedir as there was no way to initialize it.... that's when package ReadConfig arose. However later I decided to add `Init` func to Basedir so it made this change obsolete.

File: bin/percona-agent-installer/api/api.go [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+}
+
+func New(apiConnector pct.APIConnector, debug bool) *Api {
+	return &Api{
+		apiConnector: apiConnector,
+		debug:        debug,
+	}
+}
+
+func (self *Api) Init(hostname, apiKey string, headers map[string]string) (code int, err error) {
```
Comment (daniel-nichter) [inline bin/percona-agent-installer/api/api.go]: Why "self" and not "a" or something like other pkgs? (Afaik we've never used used this style before.)

Comment (arvenil) [inline bin/percona-agent-installer/api/api.go]: Because:
* I was tired of constantly changing `a` to `b` and then `c` through `func` code whenever I decided to change receiver of given func
* I was tired of constantly changing `a` to `b` and then `c` through `func` code whenever I decided to change receiver name
* I was tired of having meaningless `a`, `b` or `c` in the func doe and constantly need to check if `a` is some kind of internal variable... or... oh wait it's receiver. With `self` in the code I immediately know it is receiver.
(btw since when one char variable became the thing, I was always taught that names should be descriptive)
* I was tired of having unpredicted conflicts e.g. `i *Instance` and the for loop through index e.g. `for i := range slice {`
* And to add something with my experience/experimenting with using `self`. Since I started to use `self` in my code, writing it started to be much more smooth. Whenever I need to modify something that uses `a` I immediately hit above problems, reminding me how much I like this `self`.

The only advantage I see of having `a` instead of `self` is that it is shorter... but 1) I have auto completion, 2) I feel like understanding the code is more important than writing 3 chars more (descriptive variable names)

I understand that many articles advice to not use `self` to avoid thinking in object oriented way in golang but they also claim that package/dependency manager and dependency locking is not needed... weirdly enough it results in about ~10 package managers written by community with none working properly because of golang toolkit limitations.

If the issue here is word `self` itself then I'm fine with just `a` or `receiver` or `r` or `recv` - anything that doesn't force me to think what this `a` variable is and doesn't force me to constantly find and replace it through the code.

@mpaolino @cfsalguero @daniel-nichter Are there any other advantages/disadvantages I missed?

@mpaolino @cfsalguero @daniel-nichter  Now, I finally understand that probably you guys might want to have those `a` `b` `c` instead of `self` so if me using `self` through the code annoys you really badly then please let me know and I will try once again to adapt to using `a` `b` `c` instead of `self` :) Let me know what you think guys.

File: test/installer/my.cnf-wrong_user (hunk 3 above)
Comment (daniel-nichter) [inline test/installer/my.cnf-wrong_user]: Ok

File: bin/percona-agent-installer/api/api.go (hunk 5 above)
Comment (daniel-nichter) [inline bin/percona-agent-installer/api/api.go]: Let's discuss it with everyone at the meeting next week.

Comment (mpaolino) [inline bin/percona-agent-installer/api/api.go]: I saw it in some places and I liked it. Regarding the "object oriented thinking", I really don't see how that makes you think in a object oriented way, maybe only Python programmers might be tempted but being one I really don't see it. Lets discuss in BsAs :)

**Comment by arvenil** [comment]
@cfsalguero @daniel-nichter Any other comments? Ready to merge?

File: bin/percona-agent-installer/api/api.go (hunk 5 above)
Comment (cfsalguero) [inline bin/percona-agent-installer/api/api.go]: @arvenil We already had this conversation when I've reviewed the code and even when I know that 'self' might not be idiomatic in Go, being a Perl/Python dev, personally I like using self. For me, it makes it really easy to understand what 'self' means.

Comment (daniel-nichter) [inline bin/percona-agent-installer/api/api.go]: I'm fine with this, just 2 things:
1. We all need to commit to eventually changing all other code to match because consistency is important
1. I'd vote for `this` instead because 1) `self` is an anthropomorphism, and 2) `this` reads better like "this object's foo method/field/variable", and 3) it matches JS so even more consistency.

File: pct/basedir.go [hunk 6]
```diff
@@ -84,6 +84,14 @@ func (b *basedir) Init(path string) error {
 	return nil
 }
 
+func (b *basedir) InitExisting(path string) {
```
Comment (daniel-nichter) [inline pct/basedir.go]: This is only used in `bin/percona-agent-installer/main_test.go` so I think it should be moved from here to `test/` or that `main_test.go` specifically. Reason being: having it here is dead/unused code in agent. It makes it look like agent can init an existing basedir but it can't and shouldn't. So best to move it to avoid confusion and make it clear that this is only for testing. Also, not having this func is a matter of security: the real `Init()` does `os.Chmod(b.configDir, 0700)` for example.

File: bin/percona-agent-installer/api/api.go (hunk 5 above)
Comment (cfsalguero) [inline bin/percona-agent-installer/api/api.go]: Even with the risk of making you think I'm crazy, I would like to change my position on this.
First, I've found this good explanation about why we should avoid using self or this: http://stackoverflow.com/questions/23482068/in-go-is-naming-the-receiver-variable-self-misleading-or-good-practise/23494386#23494386

Second, there is a "wiki" with Code Review comments collected from code reviews for Go code and it says that self should be avoided (https://github.com/golang/go/wiki/CodeReviewComments#receiver-names). I know I've said I like 'self' (and I really do) but this recommendations are accepted by the community and since we have open source projects, what are we going to do if, for example, someone pushes perfect idiomatic Go code to the agent's repo? Are we going to tell him 'Sorry, we only accept using self or this'? Or are we use idiomatic Go in our open source projects and self in the api?
So, I apologize for going back and forth but I truly believe we should go back and use func (a *Api) ...

Comment (daniel-nichter) [inline bin/percona-agent-installer/api/api.go]: Indeed, https://github.com/golang/go/wiki/CodeReviewComments#receiver-names is a pretty strong argument. I still prefer *not* to use `self` or `this` because it is known not to be idiomatic Go. Not to mention all our code already *is* idiomatic. So yes, I'd rather remain idiomatic Go.

**Comment by arvenil** [comment]
@cfsalguero @daniel-nichter @mpaolino Any other comments? Ready to merge?

**Comment by cfsalguero** [comment]
:+1:

File: bin/percona-agent-installer/api/api.go [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Create new instance, if it already exist then update it
+	if resp.StatusCode == http.StatusConflict {
+		// API returns URI of existing resource in Location header
+		uri := resp.Header.Get("Location")
+		if uri == "" {
+			return nil, fmt.Errorf("API did not return location of exisiting MySQL instance")
```
Comment (mpaolino) [inline bin/percona-agent-installer/api/api.go]: s/exisiting/existing/

File: bin/percona-agent-installer/api/api.go [hunk 8]
```diff
@@ -0,0 +1,372 @@
+/*
+   Copyright (c) 2014-2015, Percona LLC and/or its affiliates. All rights reserved.
```
Comment (mpaolino) [inline bin/percona-agent-installer/api/api.go]: only 2015?

File: bin/percona-agent-installer/main_test.go [hunk 9]
```diff
@@ -138,23 +139,28 @@ func (s *MainTestSuite) TearDownTest(t *C) {
...
 
 var grantPasswordRe = regexp.MustCompile(` IDENTIFIED BY PASSWORD.+$`)
 
 func (s *MainTestSuite) GetGrants() []string {
 	grants := []string{}
-	rows, err := s.rootConn.Query("SHOW GRANTS FOR 'percona-agent'@'localhost'")
-	if err != nil {
-		fmt.Println(err)
+	rowsLocalhost, err := s.rootConn.Query("SHOW GRANTS FOR 'percona-agent'@'localhost'")
+	if val, ok := err.(*mysql.MySQLError); ok && val.Number == 1141 {
```
Comment (mpaolino) [inline bin/percona-agent-installer/main_test.go]: how about a defining a constant with the error code instead of using the magic number directly? You use it at least one more time on line 173

File: bin/percona-agent-installer/api/api.go (hunk 8 above)
Comment (arvenil) [inline bin/percona-agent-installer/api/api.go]: I don't know. It's mostly just moved code. http://stackoverflow.com/a/2391555
Let me know if I should put here 2015 or 2014-2015... legal stuff is not my point of interest.

File: bin/percona-agent-installer/api/api.go (hunk 7 above)
Comment (arvenil) [inline bin/percona-agent-installer/api/api.go]: It's an old code my friend, just moved. Fixed.

File: bin/percona-agent-installer/api/api.go (hunk 8 above)
Comment (mpaolino) [inline bin/percona-agent-installer/api/api.go]: I guess that if this was moved and written on 2014 then the year should stay as it is

**Comment by mpaolino** [comment]
:+1:

**Comment by daniel-nichter** [comment]
:+1:
